File: src/01-easy/happy-numbers/solution.cpp
#include <iostream>
#include <string>
#include <set>
using namespace std;
int dss(int n){int s=0;while(n){int d=n%10;s+=d*d;n/=10;}return s;}
int main(){
    int n; cin>>n;
    while(n--){
        string s; cin>>s;
        int x=0; for(char c:s){int d=c-'0';x+=d*d;}
        set<int> seen;
        while(x!=1&&!seen.count(x)){seen.insert(x);x=dss(x);}
        cout<<s<<(x==1?" :)":" :(")<<endl;
    }
}

File: src/01-easy/happy-numbers/solution.hs
import System.IO (hSetBuffering, stdout, BufferMode(NoBuffering))
import qualified Data.Set as S

dss :: Int -> Int
dss 0 = 0
dss n = let d = n `mod` 10 in d*d + dss (n `div` 10)

isHappy :: S.Set Int -> Int -> Bool
isHappy _ 1 = True
isHappy seen x
  | S.member x seen = False
  | otherwise = isHappy (S.insert x seen) (dss x)

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering
    n <- readLn :: IO Int
    mapM_ (\_ -> do
        s <- getLine
        let x = sum [d*d | c <- s, let d = fromEnum c - 48]
        putStrLn $ s ++ if isHappy S.empty x then " :)" else " :("
        ) [1..n]
